src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -972,15 +972,21 @@ private static int CanonicallyCompareDiagnostics(Diagnostic x, Diagnostic y)
 
             int codeCompare = xCode.CompareTo(yCode);
 
-            // ToString fails for a diagnostic with an error code that does not prevent successful delegate conversion.
-            // Also, the order doesn't matter, since all such diagnostics will be dropped.
-            if (!ErrorFacts.PreventsSuccessfulDelegateConversion(xCode) || !ErrorFacts.PreventsSuccessfulDelegateConversion(yCode))
-            {
+            // Optimization: don't bother 
+            if (codeCompare != 0)
                 return codeCompare;
+
+            for (int i = 0; i < Math.Min(x.Arguments.Count, y.Arguments.Count); i++)
+            {
+                object argx = x.Arguments[i];
+                object argy = y.Arguments[i];
+
+                codeCompare = string.CompareOrdinal(argx?.ToString(), argy?.ToString());
+                if (codeCompare != 0)
+                    return codeCompare;
             }
 
-            // Optimization: don't bother 
-            return codeCompare == 0 ? string.CompareOrdinal(x.ToString(), y.ToString()) : codeCompare;
+            return x.Arguments.Count.CompareTo(y.Arguments.Count);
```

Comment:
`return x.Arguments.Count - y.Arguments.Count`?

Reply:
That would work. Is there any advantage?